### PR TITLE
[AIRFLOW-3531] Add gcs to gcs transfer operator.

### PR DESCRIPTION
**Pull request by jmcarp**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3531
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Review by kaxil — APPROVED** [review]

File: airflow/contrib/operators/gcs_to_gcs_transfer_operator.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/transferJobs.
+        If not set, run transfer job once as soon as the operator runs
+    :type schedule: dict
+    :param object_conditions: Transfer service object conditions; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec
+    :type object_conditions: dict
+    :param transfer_options: Transfer service transfer options; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec
+    :type transfer_options: dict
+    :param wait: Wait for transfer to finish
```
Comment (kaxil) [inline airflow/contrib/operators/gcs_to_gcs_transfer_operator.py]: Add default value in description

File: airflow/contrib/operators/gcs_to_gcs_transfer_operator.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    :type description: str
+    :param schedule: Optional transfer service schedule; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/transferJobs.
+        If not set, run transfer job once as soon as the operator runs
+    :type schedule: dict
+    :param object_conditions: Transfer service object conditions; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec
+    :type object_conditions: dict
+    :param transfer_options: Transfer service transfer options; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec
```
Comment (kaxil) [inline airflow/contrib/operators/gcs_to_gcs_transfer_operator.py]: Better to point to exact location:
https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec#transferoptions

File: airflow/contrib/operators/gcs_to_gcs_transfer_operator.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        domain-wide delegation enabled.
+    :type delegate_to: str
+    :param description: Optional transfer service job description
+    :type description: str
+    :param schedule: Optional transfer service schedule; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/transferJobs.
+        If not set, run transfer job once as soon as the operator runs
+    :type schedule: dict
+    :param object_conditions: Transfer service object conditions; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec
```
Comment (kaxil) [inline airflow/contrib/operators/gcs_to_gcs_transfer_operator.py]: https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec#objectconditions

File: airflow/contrib/operators/gcs_to_gcs_transfer_operator.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    :param source_bucket: The source Google cloud storage bucket where the
+         object is. (templated)
+    :type source_bucket: str
+    :param destination_bucket: The destination Google cloud storage bucket
+        where the object should be. (templated)
+    :type destination_bucket: str
+    :param project_id: The ID of the Google Cloud Platform Console project that
+        owns the job
+    :type project_id: str
+    :param gcp_conn_id: The destination connection ID to use
```
Comment (kaxil) [inline airflow/contrib/operators/gcs_to_gcs_transfer_operator.py]: You still need read on the source bucket, right?

**Review by kaxil — CHANGES_REQUESTED** [review]

**Comment by kaxil** [comment]
Any updates on this PR @jmcarp

**Comment by jmcarp** [comment]
Updated!

File: airflow/contrib/operators/gcs_to_gcs_transfer_operator.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        For this to work, the service account making the request must have
+        domain-wide delegation enabled.
+    :type delegate_to: str
+    :param description: Optional transfer service job description
+    :type description: str
+    :param schedule: Optional transfer service schedule; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/transferJobs.
+        If not set, run transfer job once as soon as the operator runs
+    :type schedule: dict
+    :param object_conditions: Transfer service object conditions; see
```
Comment (kaxil) [inline airflow/contrib/operators/gcs_to_gcs_transfer_operator.py]: Can you just add info that it is Optional param

File: airflow/contrib/operators/gcs_to_gcs_transfer_operator.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec#ObjectConditions
+    :type object_conditions: dict
+    :param transfer_options: Transfer service transfer options; see
```
Comment (kaxil) [inline airflow/contrib/operators/gcs_to_gcs_transfer_operator.py]: Can you just add info that it is Optional param

File: airflow/contrib/operators/gcs_to_gcs_transfer_operator.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    """
+
+    template_fields = ('source_bucket', 'destination_bucket', 'description', 'object_conditions')
+    ui_color = '#e09411'
+
+    @apply_defaults
+    def __init__(self,
+                 source_bucket,
+                 destination_bucket,
+                 project_id,
```
Comment (kaxil) [inline airflow/contrib/operators/gcs_to_gcs_transfer_operator.py]: `project_id ` should be optional as we have it from the Connections

File: docs/integration.rst [hunk 8]
```diff
@@ -1184,6 +1184,7 @@ Storage Operators
 - :ref:`GoogleCloudStorageObjectCreateAclEntryOperator` : Creates a new ACL entry on the specified object.
 - :ref:`GoogleCloudStorageToBigQueryOperator` : Loads files from Google cloud storage into BigQuery.
 - :ref:`GoogleCloudStorageToGoogleCloudStorageOperator` : Copies objects from a bucket to another, with renaming if requested.
+- :ref:`GoogleCloudStorageToGoogleCloudStorageTransferOperator` : Copies objects from a bucket to another, with renaming if requested using the GCP Storage Transfer Service.
```
Comment (kaxil) [inline docs/integration.rst]: It doesn't allow renaming, does it?

**Review by kaxil — CHANGES_REQUESTED** [review]

**Comment by jmcarp** [comment]
Updated.

**Comment by jmcarp** [comment]
Ready for another look when you have time @kaxil

**Review by kaxil — APPROVED** [review]

**Pull request merged**